[2D] Bind _fsdp_extension to FSDP instances

wz337 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113237

Currently, when we have 2D composition, a global variable _extensions controls the 2D deviation we need to take in state_dict calls (See https://github.com/pytorch/pytorch/blob/release/2.1/torch/distributed/fsdp/_fsdp_extensions.py#L66-L68). This is problematic when we have both a 2D model and a plain FSDP model in the same dist environment, as the _extensions will be mistakenly turned on for the plain FSDP model, resulting in state_dict error (RuntimeError: No parent device_mesh is found for FSDP device_mesh.).

This PR binds _fsdp_extension to the FSDP instances to make sure that state_dict calls would not get interfered with each other when mixing both 2D and 1D parallelism.

wz337 committed:
bind extension to a FSDP instance

[ghstack-poisoned]

wz337 committed:
Update on "[2D] Bind _fsdp_extension to FSDP instances"


Currently, when we have 2D composition, a global variable _extensions controls the 2D deviation we need to take in state_dict calls (See https://github.com/pytorch/pytorch/blob/release/2.1/torch/distributed/fsdp/_fsdp_extensions.py#L66-L68). This is problematic when we have both a 2D model and a plain FSDP model in the same dist environment, as the _extensions will be mistakenly turned on for the plain FSDP model, resulting in state_dict error (RuntimeError: No parent device_mesh is found for FSDP device_mesh.).

This PR binds _fsdp_extension to the FSDP instances to make sure that state_dict calls would not get interfered with each other when mixing both 2D and 1D parallelism.


[ghstack-poisoned]

fduwjj commented on torch/distributed/fsdp/_shard_utils.py:
```diff
@@ -15,7 +15,6 @@
 )
 from torch.distributed._shard.sharding_spec import ShardMetadata
 from torch.distributed._tensor import DeviceMesh, DTensor, Replicate, Shard as DShard
-from torch.distributed.fsdp._debug_utils import SimpleProfiler
```
nit: maybe you want to split this change into a separate PR? (Maybe in the future)

wz337 replied:
This actually needs to go in the same PR as it was resulting in a circular dependency without this, since now we need to add the DTensorExtension to the FSDPstate.

fduwjj commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -512,20 +511,12 @@ def _init_prefetching_state(
-                    "2D parallelism won't work with FSDP"
-                    f"exception: {e}"
-                )
+    if device_mesh and _mesh_resources.get_parent_mesh(state._device_mesh) is not None:
+        state._fsdp_extension = DTensorExtensions()
+    else:
+        # We need to explicilty set _fsdp_extension to None.
+        # Otherwise, we will run into an infinite recursion when getting the attribute.
+        state._fsdp_extension = None
```
Nice nice!

fduwjj commented on torch/distributed/fsdp/_common_utils.py:
```diff
@@ -1,7 +1,6 @@
 """
 This file includes private common utilities for FSDP.
 """
-
```
is this expected?

fduwjj commented:
Also please make linter happy.

wz337 committed:
Update on "[2D] Bind _fsdp_extension to FSDP instances"


Currently, when we have 2D composition, a global variable _extensions controls the 2D deviation we need to take in state_dict calls (See https://github.com/pytorch/pytorch/blob/release/2.1/torch/distributed/fsdp/_fsdp_extensions.py#L66-L68). This is problematic when we have both a 2D model and a plain FSDP model in the same dist environment, as the _extensions will be mistakenly turned on for the plain FSDP model, resulting in state_dict error (RuntimeError: No parent device_mesh is found for FSDP device_mesh.).

This PR binds _fsdp_extension to the FSDP instances to make sure that state_dict calls would not get interfered with each other when mixing both 2D and 1D parallelism.


[ghstack-poisoned]

wz337 committed:
Update on "[2D] Bind _fsdp_extension to FSDP instances"


Currently, when we have 2D composition, a global variable _extensions controls the 2D deviation we need to take in state_dict calls (See https://github.com/pytorch/pytorch/blob/release/2.1/torch/distributed/fsdp/_fsdp_extensions.py#L66-L68). This is problematic when we have both a 2D model and a plain FSDP model in the same dist environment, as the _extensions will be mistakenly turned on for the plain FSDP model, resulting in state_dict error (RuntimeError: No parent device_mesh is found for FSDP device_mesh.).

This PR binds _fsdp_extension to the FSDP instances to make sure that state_dict calls would not get interfered with each other when mixing both 2D and 1D parallelism.


[ghstack-poisoned]

fegin reviewed (commented):
After this PR, what's the plan to let users set the customized extension?

fegin reviewed (approved):
https://github.com/pytorch/pytorch/pull/112835 is landed. You can rebase to avoid the change of _shard_utils.py.

wz337 committed:
Update on "[2D] Bind _fsdp_extension to FSDP instances"


Currently, when we have 2D composition, a global variable _extensions controls the 2D deviation we need to take in state_dict calls (See https://github.com/pytorch/pytorch/blob/release/2.1/torch/distributed/fsdp/_fsdp_extensions.py#L66-L68). This is problematic when we have both a 2D model and a plain FSDP model in the same dist environment, as the _extensions will be mistakenly turned on for the plain FSDP model, resulting in state_dict error (RuntimeError: No parent device_mesh is found for FSDP device_mesh.).

This PR binds _fsdp_extension to the FSDP instances to make sure that state_dict calls would not get interfered with each other when mixing both 2D and 1D parallelism.


[ghstack-poisoned]

wz337 commented:
@pytorchmergebot merge
